Sources/PackageCollections/PackageIndex+Configuration.swift

```diff
@@ -0,0 +1,149 @@
+        self.cacheTTLInSeconds = cacheTTLInSeconds ?? 3600
+        self.cacheSizeInMegabytes = cacheSizeInMegabytes ?? 10
+    }
+}
+
+public struct PackageIndexConfigurationStorage {
+    private let path: AbsolutePath
+    private let fileSystem: FileSystem
+
+    private let encoder: JSONEncoder
+    private let decoder: JSONDecoder
+
+    public init(path: AbsolutePath, fileSystem: FileSystem) {
+        self.path = path
+        self.fileSystem = fileSystem
+        self.encoder = JSONEncoder.makeWithDefaults()
+        self.decoder = JSONDecoder.makeWithDefaults()
+    }
+    
+    public func load() throws -> PackageIndexConfiguration {
+        guard self.fileSystem.exists(self.path) else {
+            return .init()
+        }
+        let buffer = try self.fileSystem.readFileContents(self.path).contents
+        guard buffer.count > 0 else {
+            return .init()
+        }
+        let container = try decoder.decode(StorageModel.Container.self, from: Data(buffer))
+        return try PackageIndexConfiguration(container.index)
+    }
+
+    public func save(_ configuration: PackageIndexConfiguration) throws {
+        if !self.fileSystem.exists(self.path.parentDirectory) {
+            try self.fileSystem.createDirectory(self.path.parentDirectory, recursive: true)
+        }
+        let container = StorageModel.Container(configuration)
+        let buffer = try encoder.encode(container)
+        try self.fileSystem.writeFileContents(self.path, bytes: ByteString(buffer), atomically: true)
+    }
+    
+    @discardableResult
+    public func update(with handler: (inout PackageIndexConfiguration) throws -> Void) throws -> PackageIndexConfiguration {
+        let configuration = try self.load()
+        var updatedConfiguration = configuration
+        try handler(&updatedConfiguration)
+        if updatedConfiguration != configuration {
+            try self.save(updatedConfiguration)
+        }
+        return updatedConfiguration
+    }
+}
+
+private enum StorageModel {
+    struct Container: Codable {
+        var index: Index
+
+        init(_ from: PackageIndexConfiguration) {
+            self.index = .init(
+                url: from.url?.absoluteString,
+                searchResultMaxItemsCount: from.searchResultMaxItemsCount,
+                cacheDirectory: from.cacheDirectory.pathString,
+                cacheTTLInSeconds: from.cacheTTLInSeconds,
+                cacheSizeInMegabytes: from.cacheSizeInMegabytes
+            )
+        }
+    }
+
+    struct Index: Codable {
+        let url: String?
+        let searchResultMaxItemsCount: Int?
+        let cacheDirectory: String?
+        let cacheTTLInSeconds: Int?
+        let cacheSizeInMegabytes: Int?
+    }
+}
+
+// MARK: - Utility
+
+private extension PackageIndexConfiguration {
+    init(_ from: StorageModel.Index) throws {
+        let url: Foundation.URL?
+        switch from.url {
+        case .none:
+            url = nil
```

Comment:
a bit surprised that URL is optional in this context, but I assume there are reasons when this makes sense?

Reply:
This is to handle the case when there is no configuration at all or when user wants to disable the feature.